### PR TITLE
Improve memory usage / performance when sending DICOM files over the network

### PR DESCRIPTION
**Pull request by amoerie**
#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Remove IByteBuffer.GetByteRange overload that returns a byte array. You now always have to pass in a buffer.
- Use shared arraypool when constructing the preamble of a PDU
- Add test coverage for various implementations of IByteBuffer

These changes are a subset of pull request #1263, I have picked the most conservative commits that focus on memory usage / performance when sending data over the network

### CODE REVIEW

**Comment by amoerie** [comment]
@gofal

This PR contains the most conservative improvements of my memory pull request that I am very comfortable with, and is ready for review.

**Comment by amoerie** [comment]
Small update: I'm marking this as draft until I've had time to do some more real world testing, I want to avoid introducing similar issues to #1264

**Comment by amoerie** [comment]
@gofal I am commenting here to address your concerns with regards to my changes in EvenLengthBuffer.

So, to recap, this is the original implementation:

EvenLengthBuffer (original)
```csharp
        public void CopyToStream(Stream s, long offset, int count)
        {
            s.Write(Buffer.Data, (int)offset, (int)Math.Min(Buffer.Size - offset, count));
            if (count > Buffer.Size - offset)
            {
                for (var i = 0; i < count - (Buffer.Size - offset); i++)
                {
                    s.WriteByte(0);
                }
            }
        }
```

EvenLengthBuffer (#1267)
```csharp
        public void CopyToStream(Stream s)
        {
            s.Write(Buffer.Data, 0, (int)Buffer.Size);
            s.WriteByte(0);
        }
``` 

The first thing you should notice is that the method signature is simplified now: there is no offset or count parameter anymore. We always write the *full* buffer to the destination stream.

The second thing you should know is that EvenLengthBuffer has a private constructor, and is always instantiated with this static factory method:

```csharp
        /// <summary>
        /// If necessary, creates an even length buffer for the specified <paramref name="buffer"/>.
        /// </summary>
        /// <param name="buffer">Buffer that is required to be of even length.</param>
        /// <returns>
        /// If <paramref name="buffer"/> is of uneven length, returns an even length buffer wrapping the <paramref name="buffer"/>,
        /// otherwise returns the buffer itself.
        /// </returns>
        public static IByteBuffer Create(IByteBuffer buffer)
            => (buffer.Size & 1) == 1
            ? new EvenLengthBuffer(buffer)
            : buffer;
```

So, to summarize:

1. We only create an EvenLengthBuffer is the inner Buffer is uneven
2. We always write the full buffer to the destination stream
3. Taking (1) and (2) together, copying an EvenLengthBuffer to a destination stream is as simple as writing the full buffer and appending a single 0 byte at the end. This ensures the output is even.

**Review by gofal — APPROVED** [review]
👍

File: FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs [hunk 1]
```diff
@@ -46,7 +47,21 @@ public CompositeByteBuffer(params IByteBuffer[] buffers)
...
+                {
+                    if (!b.IsMemory)
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+        }
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs:65]:
```cs
public bool IsMemory => Buffers.All(b => b.IsMemory);
```
Or is that slower?

File: FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs [hunk 2]
```diff
@@ -110,8 +116,17 @@ public void GetByteRange(long offset, int count, byte[] output)
...
+                    try
+                    {
+                        Buffers[pos].GetByteRange(offset, remain, temp);
+                        
+                        System.Buffer.BlockCopy(temp, 0, output, offset2, remain);
+                    }
+                    finally
+                    {
+                        ArrayPool<byte>.Shared.Return(temp);
+                    }
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs:129]: As this is used so often, maybe it would make sense to add a context manager, e.g. something like:
```cs
    public class BytePool : IDisposable
    {
        public byte[] Buffer { get; }

        public BytePool(int bufferSize)
        {
            Buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
        }

        public void Dispose() => ArrayPool<byte>.Shared.Return(Buffer);
    }
```
and use it like 
```cs
    using var pool = new BytePool(remain);
    Buffers[pos].GetByteRange(offset, remain, pool.Buffer);
    System.Buffer.BlockCopy(pool.Buffer, 0, output, offset2, remain);
```

File: FO-DICOM.Core/Network/PDU.cs [hunk 3]
```diff
@@ -99,11 +101,23 @@ public RawPDU(MemoryStream stream, Encoding encoding = null)
...
+                GetCommonFields(preamble);
+
+                s.Write(preamble, 0, 6);
+                
+                _ms.Seek(0, SeekOrigin.Begin);
+                
+                _ms.CopyTo(s);
+                
+                s.Flush();
+            }
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/Network/PDU.cs:116]: Just out of curiosity: is there a reason for all these newlines between the statements? Is it some grouping?

**Review by mrbean-bremen — APPROVED** [review]
This looks quite good to me, as far as I understand the code, had only a couple of questions/remarks. The test coverage for that code increases sufficiently (though admittetly I only scanned the tests, didn't review them in depth).

File: FO-DICOM.Core/DicomEncoding.cs [hunk 4]
```diff
@@ -287,13 +306,13 @@ internal static string DecodeBytes(IByteBuffer buffer, Encoding[] encodings, boo
             936 // gb2312
         };
 
-        private static string DecodeFragment(byte[] fragment, Encoding[] encodings, byte[] delimiters)
+        private static string DecodeFragment(byte[] fragment, int fragmentLength, Encoding[] encodings, byte[] delimiters)
```
[the remark this answers is not in the document]
Comment (amoerie) [reply FO-DICOM.Core/DicomEncoding.cs:309]: The length a rented byte array is usually longer than what you requested, so you need to manually ensure you always use only the requested length part of a rented array.

File: FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs (hunk 2 above)
Comment (amoerie) [reply FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs:129]: Good suggestion! I've been meaning to do this, as it would also enable us to implement a finalizer that could act as a safety net against forgetting to dispose or against double disposal. I intend to implement this in the next batch of memory improvements but I didn't want to put too many changes into a single pull request like last time...

File: FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs (hunk 1 above)
Comment (amoerie) [reply FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs:65]: Hm the only way in which it is slower is that you have a function invocation for each item, but that overhead should be minimal. I guess brevity wins here.

File: FO-DICOM.Core/Network/PDU.cs (hunk 3 above)
Comment (amoerie) [reply FO-DICOM.Core/Network/PDU.cs:116]: Ah no, just the result of me changing my mind a lot and refactoring. I'll clean it up a bit.

File: FO-DICOM.Core/DicomEncoding.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrbean-bremen) [reply FO-DICOM.Core/DicomEncoding.cs:309]: Ah ok, thanks!

File: FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs (hunk 2 above)
Comment (amoerie) [reply FO-DICOM.Core/IO/Buffer/CompositeByteBuffer.cs:129]: If it's alright with you @mrbean-bremen I'll resolve this conversation, you can poke me about this if I forget in the next PR.
